# scarHRD

scarHRD is an R package which determines the levels of homologous recombination deficiency (telomeric allelic imbalance, loss of heterozygosity, number of large-scale transitions) based on NGS (WES, WGS) data.

## Citation

Please cite the following paper: manuscript submitted.


## Installation

scarHRD is an R package, and can be installed via devtools.

```{r}
devtools::install_github('sztup/scarHRD')
```

## Usage example

```{r}
library(scarHRD)
scar_score("F:/test1.small.seqz.gz",reference = "grch38", seqz=TRUE)
scar_score("F:/test2.txt",reference = "grch38", seqz=FALSE)
```

Where the the input file may be a binned seqz file (small.seqz) generated by sequenza (or any other segmentation file with the following columns: chromosome, position, base.ref, depth.normal, depth.tumor, depth.ratio, Af, Bf, zygosity.normal, GC.percent, good.reads, AB.normal, AB.tumor, tumor.strand), or an allele-specific segmentation file with the following columns: 1st column: sample name, 2nd column: chromosome, 3rd column: segmentation start, 4th column: segmentation end, 5th column: total copynumber, 6th column: copy number of A allele, 7th column: copy number of B allele.
